#!/bin/bash
#PBS -V
#PBS -o $OUTFILE.$MOAB_JOBARRAYINDEX.log

cd $PBS_O_WORKDIR

. $LIB_DIR/load_opts $CONFIG

#reads from: SPLIT_FQ
#working files: CLIPFILE ALNFILE TMPSAM
#writes to: TMPBAM

SECONDS=0;

log_info "Clipping file $SPLIT_FQ to $CLIPFILE"
# Clip out the adapter sequences from the fastq file
zcat $SPLIT_FQ | p_fastx_clipper -a $FASTX_ADAPTER -l $FASTX_MINLENGTH $FASTX_QUALS -n -v -z -o $CLIPFILE || { error_exit "fastx_clipper failed."; }
# Align against the BWA index
log_info "Aligning file $CLIPFILE to $ALNFILE"
p_bwa aln -t $BWA_THREADS -f $ALNFILE $BWA_IDX $CLIPFILE || { error_exit "BWA aln failed."; }
# Convert single-end reads to genomic coordinates (could be adapted to do paired-ends)
log_info "Mapping coordinates to genome from $ALNFILE"
p_bwa samse -f $TMPSAM $BWA_IDX $ALNFILE $CLIPFILE || { error_exit "BWA samse failed."; }
# Swap out the coordinates
log_info "Building BAM file $TMPBAM"
p_picard AddOrReplaceReadGroups SO=coordinate CREATE_INDEX=true I=$TMPSAM O=$TMPBAM ID=$SCRATCH LB=$SCRATCH PL=illumina PU=None SM=$SCRATCH CN=Babel@UCSF || { error_exit "Picard failed."; }

rm -f $CLIPFILE
rm -f $ALNFILE
rm -f $TMPSAM
