#!/bin/bash
#
# Perform a known junction-only alignment with Tophat of those
# reads that did not align to the genome with BWA.
# (do so on nodes n012:ppn=24 or n013:ppn=24
#
#PBS -V
#PBS -o $OUTFILE.log

#reads from: UNALIGNED_FASTQ
#working files: TOPHAT_SCRATCH/*
#writes to: TOPHAT_BAM

cd $PBS_O_WORKDIR

. $LIB_DIR/load_opts $CONFIG

log_info "Aligning unaligned reads from $UNALIGNED_FASTQ"
log_info "Using genome index $BOWTIE_IDX"
log_info "Using transcriptome index $TRANSCRIPTOME_IDX"
# Run tophat on unaligned reads
p_tophat -p $TOPHAT_THREADS \
	-g $TOPHAT_MULTIHITS \
	-a $TOPHAT_MINANCHORLENGTH \
	-m $TOPHAT_MISMATCHES \
	$TOPHAT_OPTS \
	$TOPHAT_QUALS \
	-o $TOPHAT_SCRATCH \
	--transcriptome-index=$TRANSCRIPTOME_IDX \
	$BOWTIE_IDX $UNALIGNED_FASTQ || { error_exit "Tophat failed"; }

mv $TOPHAT_SCRATCH/unmapped_left.fq.z $TOPHAT_SCRATCH/unmapped_left.fq.gz || { error_exit "Rename failed"; }
# Save the reads that Tophat couldn't align
p_picard FastqToSam FASTQ=$TOPHAT_SCRATCH/unmapped_left.fq.gz QUALITY_FORMAT=Standard OUTPUT=$TOPHAT_SCRATCH/unaligned.bam SAMPLE_NAME=$BASE
# Merge them back in with the rest
p_picard MergeSamFiles O=$TOPHAT_SCRATCH/merged.bam I=$TOPHAT_SCRATCH/accepted_hits.bam I=$TOPHAT_SCRATCH/unaligned.bam MSD=true
# Label the merged groups
log_info "Creating temp BAM $TOPHAT_BAM"
p_picard AddOrReplaceReadGroups CREATE_INDEX=true SO=coordinate I=$TOPHAT_SCRATCH/merged.bam O=$TOPHAT_BAM ID=$BASE LB=$BASE PL=illumina PU=None SM=$BASE CN=Babel@UCSF
log_output "Writing metrics to $TOPHATFLAG_FILE"
# Get stats for the Tophat results
samtools flagstat $TOPHAT_BAM > $TOPHATFLAG_FILE

rm -rf $TOPHAT_SCRATCH
log_info "Done!"
